internal/provider/data_source_docker_image.go

```diff
@@ -0,0 +1,62 @@
+package provider
+
+import (
+	"context"
+	"log"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func dataSourceDockerImage() *schema.Resource {
+	return &schema.Resource{
+		Description: "`docker_image` provides details about a specific Docker Image which need to be presend on the Docker Host",
+
+		ReadContext: dataSourceDockerImageRead,
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:        schema.TypeString,
+				Description: "The name of the Docker image, including any tags or SHA256 repo digests.",
+				Required:    true,
+			},
+			"latest": {
+				Type:        schema.TypeString,
+				Description: "The ID of the image in the form of `sha256:<hash>` image digest. Do not confuse it with the default `latest` tag.",
+				Computed:    true,
+				Deprecated:  "Use `sha256_digest` instead",
+			},
+			"sha256_digest": {
+				Type:        schema.TypeString,
+				Description: "The image sha256 digest in the form of `sha256:<hash>`.",
+				Computed:    true,
+			},
+		},
+	}
+}
+
+func dataSourceDockerImageRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*ProviderConfig).DockerClient
+
+	var data Data
+	if err := fetchLocalImages(ctx, &data, client); err != nil {
+		return diag.Errorf("Error reading docker image list: %s", err)
+	}
+	for id := range data.DockerImages {
+		log.Printf("[DEBUG] local images data: %v", id)
+	}
+
+	imageName := d.Get("name").(string)
+
+	foundImage := searchLocalImages(ctx, client, data, imageName)
+	if foundImage == nil {
+		return diag.Errorf("did not find docker image '%s'", imageName)
+	}
+
+	d.SetId(foundImage.ID)
+	d.Set("name", imageName)
+	d.Set("latest", foundImage.ID)
+	d.Set("sha256_digest", foundImage.ID)
```

Comment:
This is not an image digest but an image id.

Reply:
I guess ImageSummary doesn't have an image digest although I don't check.
https://pkg.go.dev/github.com/docker/docker@v20.10.6+incompatible/api/types#ImageSummary

Reply:
This is a good point -- while it may appear to be a sha256 digest (and maybe that's how it's generated from some collection of uniquely identifying data), it's still simply an `id`.

If `d.SetId(foundImage.ID)` also sets the `id` attribute, why not use that instead of `latest` or `sha256_digest` altogether, if it does indeed uniquely identify the image?

Reply:
If we have (or can construct) the `repo_digest`, that'd be the best thing to have here -- especially since `docker_service` will need to pass all that information along to the nodes it gets scheduled on, so they know what to pull.
```
m1:~$ docker image inspect 075d847609aa
[
    {
        "Id": "sha256:075d847609aa84dbbe8a5a36ca3852dd75bbcd643a92bb5e1757e75727d44210",
        "RepoTags": [
            "jakesys/aws2:latest"
        ],
        "RepoDigests": [
            "jakesys/aws2@sha256:aa579322e9a0a27c5bece3cc9f5a76e5482c2c846e555917f17ce28034a78d3a"
        ],
...
}

Reply:
Good point. What confused me was that `RepoDigests` is an array. So what happens if there a multiple? Shall we take the first?

Reply:
Taking a closer look at what multiple `RepoDigests` means, I've managed to get multiple values...
```
jake@jimini ~ % docker image ls                                                     
REPOSITORY                                TAG               IMAGE ID       CREATED        SIZE
jakesys/aws2                              latest            a406618724ac   10 hours ago   95.7MB
registry.gitlab.com/jakesys/aws2/master   2.2.9             a406618724ac   10 hours ago   95.7MB
registry.gitlab.com/jakesys/aws2/master   latest            a406618724ac   10 hours ago   95.7MB
jakesys/aws2                              <none>            7f92c8cc54e5   2 days ago     95.7MB
```
...because the same image can be in _multiple_ registries.  The untagged image was formerly the `latest`, the new `latest` was built by GitLab CI last night, and pushed to hub.docker.com.  I also pulled the corresponding version tag from the GitLab registry.
```
jake@jimini ~ % docker image inspect jakesys/aws2
[
    {
        "Id": "sha256:a406618724ac720e08c077b604c25946f6c5da1702f5e0b30f72c91d1e40643f",
        "RepoTags": [
            "jakesys/aws2:latest",
            "registry.gitlab.com/jakesys/aws2/master:2.2.9",
            "registry.gitlab.com/jakesys/aws2/master:latest"
        ],
        "RepoDigests": [
            "jakesys/aws2@sha256:3734f6bf3c8bd2a384caba078a4e4913f8809f871b473978bbf27cf726977b1c",
            "registry.gitlab.com/jakesys/aws2/master@sha256:3734f6bf3c8bd2a384caba078a4e4913f8809f871b473978bbf27cf726977b1c"
        ],
...
  }
]
```

The `RepoDigest` that matches the image name being used to define the `docker_registry_image` data source (or the Terraform-managed resource) would be the one to use.

Reply:
Looks good.
The the fallback should be documented.

Reply:
I'm going to disagree here.  If the local `docker_image` data source / resource does not have these values set, a `null` value would be more appropriate to indicate that the image was not pulled from or pushed to a registry.  Also the image's `ID` string has a completely different format than the `repoDigest`.

We shouldn't need to fill in all the empty blanks...  ;)

Certainly we should document what could cause a `null` (or empty string) value for the read-only attribute.

Fleshing out the `repo_digest` attribute on the `docker_registry_image` data source / resource shouldn't have this issue... :)

Reply:
See d04ff80e1808dd18adfaf3a9354947a1853d3759 and the docs in 5ec1577bdbba9cbce5ae13812a219741013b4628

Reply:
this is better -- but there's still one more edge case where the image was pulled from somewhere, pushed somewhere else, but the tag being referenced in the `data` is that local-only tag...

i think the following would cover all cases...
* empty `RepoDigests`, return `""`
* one `RepoDigests`, return it
* multiple `RepoDigests`...
  * one matches the `imageName`, return it
  * nothing matches the `imageName`, return the first one *(new edge case)*

That should cover all situations.

Reply:
Alright, I added a fix in 2c5d6f8e48a805cfa38420de3fce594466fecc6f and did throw an error if _nothing matches_

Reply:
Looks good, thanks!